lib/backend/k8s/conversion.go

```diff
@@ -0,0 +1,359 @@
+	// Parse out important fields.
+	policyName := fmt.Sprintf("%s.%s", np.ObjectMeta.Namespace, np.ObjectMeta.Name)
+	order := float64(1000.0)
+
+	// Generate the inbound rules list.
+	inboundRules := []model.Rule{}
+	for _, r := range np.Spec.Ingress {
+		inboundRules = append(inboundRules, c.parseIngressRule(r, np.ObjectMeta.Namespace)...)
+	}
+
+	// Build and return the KVPair.
+	return &model.KVPair{
+		Key: model.PolicyKey{
+			Name: policyName,
+		},
+		Value: &model.Policy{
+			Order:         &order,
+			Selector:      c.parseSelector(&np.Spec.PodSelector, &np.ObjectMeta.Namespace),
+			InboundRules:  inboundRules,
+			OutboundRules: []model.Rule{},
+		},
+		Revision: np.ObjectMeta.ResourceVersion,
+	}, nil
+}
+
+// parseSelector takes a namespaced k8s label selector and returns the Calico
+// equivalent.
+func (c converter) parseSelector(s *unversioned.LabelSelector, ns *string) string {
+	// If this is a podSelector, it needs to be namespaced, and it
+	// uses a different prefix.  Otherwise, treat this as a NamespaceSelector.
+	selectors := []string{}
+	prefix := "k8s_ns/label/"
+	if ns != nil {
+		prefix = ""
+		selectors = append(selectors, fmt.Sprintf("calico/k8s_ns == '%s'", *ns))
+	}
+
+	// matchLabels is a map key => value, it means match if (label[key] ==
+	// value) for all keys.
+	for k, v := range s.MatchLabels {
+		selectors = append(selectors, fmt.Sprintf("%s%s == '%s'", prefix, k, v))
+	}
+
+	// matchExpressions is a list of in/notin/exists/doesnotexist tests.
+	for _, e := range s.MatchExpressions {
+		valueList := strings.Join(e.Values, ", ")
+
+		// Each selector is formatted differently based on the operator.
+		switch e.Operator {
+		case unversioned.LabelSelectorOpIn:
+			selectors = append(selectors, fmt.Sprintf("%s%s in { %s }", prefix, e.Key, valueList))
+		case unversioned.LabelSelectorOpNotIn:
+			selectors = append(selectors, fmt.Sprintf("%s%s no int { %s }", prefix, e.Key, valueList))
+		case unversioned.LabelSelectorOpExists:
+			selectors = append(selectors, fmt.Sprintf("has(%s%s)", prefix, e.Key))
+		case unversioned.LabelSelectorOpDoesNotExist:
+			selectors = append(selectors, fmt.Sprintf("! has(%s%s)", prefix, e.Key))
+		}
+	}
+
+	return strings.Join(selectors, " && ")
+}
+
+func (c converter) parseIngressRule(r extensions.NetworkPolicyIngressRule, ns string) []model.Rule {
+	rules := []model.Rule{}
+	peers := []*extensions.NetworkPolicyPeer{}
+	ports := []*extensions.NetworkPolicyPort{}
+
+	// Built up a list of the sources and a list of the desintations.
+	for _, f := range r.From {
+		peers = append(peers, &f)
+	}
+	for _, p := range r.Ports {
+		ports = append(ports, &p)
+	}
+
+	// If there no peers, or no ports, represent that as nil.
+	if len(peers) == 0 {
+		peers = []*extensions.NetworkPolicyPeer{nil}
+	}
+	if len(ports) == 0 {
+		ports = []*extensions.NetworkPolicyPort{nil}
+	}
+
+	// Combine desintations with sources to generate rules.
+	for _, port := range ports {
```

Comment:
I'm going to leave this as-is for now, since it seems to work and we're a bit crunched for time.  Let me know if you think there's a reason I should do it now instead.